### DISCUSSION TITLE
ifEmpty()->thenUnset() not working for arrayNode() ?

### DISCUSSION THREAD

**Discussion by wtfred** [symfony/symfony#43249] [Q&A] [answered]
Hello,
I'm using the config component and having a bit of trouble. Here is what i'm doing, and what i'm trying to achieve.
With this in my configuration class
```
$node
    ->children()
        ->scalarNode('title')
            ->isRequired()
            ->cannotBeEmpty()
        ->end()
        ->arrayNode('drivers')
            ->scalarPrototype()->end()
        ->end()
    ->end()
;
```
i can have a yaml file like this
```
parent:
  title: "test"
  drivers:
    - value1
    - value2
```
And after use of the Processor, i end up with this array
```
parent => array(2 items)
  title => "test" (4 chars)
  drivers => array(2 items)
    0 => "value1" (6 chars)
    1 => "value2" (6 chars)
```
Until now, everything is working fine. But now, i want the `drivers` part from yaml to be optional, so i removed it
```
parent:
  title: "test"
```
Then i have this:
```
parent => array(2 items)
  title => "test" (4 chars)
  drivers => array(empty)
```
Ok then, arrayNode have a default value, which is an empty array. Fine, but i don't want it. So i tried to remove it in my treeBuilder
```
$node
    ->children()
        ->scalarNode('title')
            ->isRequired()
            ->cannotBeEmpty()
        ->end()
        ->arrayNode('drivers')
            ->validate()
                ->ifEmpty()->thenUnset()
            ->end()
            ->scalarPrototype()->end()
        ->end()
    ->end()
;
```
But this has no effect, i always end up with
```
parent => array(2 items)
  title => "test" (4 chars)
  drivers => array(empty)
```
Did i miss something? Is there a way to achieve what i want? Is this a bug or normal behaviour?
Only "solution" i found on stackoverflow is to add this on parent validation, and i really don't like it ^^
```
->validate()
    ->always()
    ->then(function ($v) {
        if(empty($v['drivers'])){
            unset($v['drivers']);
        }
        return $v;
    })
->end()
```

Thank you guys for reading, hope you can help me find a solution :)

**Comment by stof** [comment, marked as answer]
prototyped array nodes always default to an empty array. You cannot unset default values

**Reply by wtfred** [reply]
Ok thanks for your answer, but is my use case considered marginal, and i have to use the "not dynamic" validate on parent solution, or is it a legitimate and maybe the component could evolve in the future to handle it if i create an issue?

**Reply by wtfred** [reply]
I'm marking this as answser, but i'm really interested in knowing if my use case is that marginal ^^
Anyway, thanks for answering :+1:
